provider/oci/networking.go

```diff
@@ -4,65 +4,981 @@
+		5*time.Minute)
+	if err != nil {
+		return ociCore.SecurityList{}, errors.Trace(err)
+	}
+	return response.SecurityList, nil
+}
+
+func (e *Environ) allSubnets(controllerUUID string, vcnID *string) (map[string][]ociCore.Subnet, error) {
+	request := ociCore.ListSubnetsRequest{
+		CompartmentId: e.ecfg().compartmentID(),
+		VcnId:         vcnID,
+	}
+	response, err := e.Networking.ListSubnets(context.Background(), request)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := map[string][]ociCore.Subnet{}
+	for _, val := range response.Items {
+		if tag, ok := val.FreeformTags[tags.JujuController]; ok {
+			if tag == controllerUUID {
+				cidr := *val.CidrBlock
+				if valid, err := e.validateCidrBlock(cidr); err != nil || !valid {
+					logger.Warningf("failed to validate CIDR block %s: %s", cidr, err)
+					continue
+				}
+				ret[*val.AvailabilityDomain] = append(ret[*val.AvailabilityDomain], val)
+			}
+		}
+	}
+	return ret, nil
+}
+
+func (e *Environ) validateCidrBlock(cidr string) (bool, error) {
+	_, vncIPNet, err := net.ParseCIDR(DefaultAddressSpace)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	subnetIP, _, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if vncIPNet.Contains(subnetIP) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (e *Environ) getFreeSubnet(existing map[string]bool) (string, error) {
+	ip, _, err := net.ParseCIDR(DefaultAddressSpace)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	to4 := ip.To4()
+	if to4 == nil {
+		return "", errors.Errorf("invalid IPv4 address: %s", DefaultAddressSpace)
+	}
+
+	for i := 0; i <= 255; i++ {
+		to4[2] = byte(i)
+		subnet := fmt.Sprintf("%s/%s", to4.String(), SubnetPrefixLength)
+		if _, ok := existing[subnet]; ok {
+			continue
+		}
+		existing[subnet] = true
+		return subnet, nil
+	}
+	return "", errors.Errorf("failed to find a free subnet")
+}
+
+func (e *Environ) getSubnetStatus(resourceID *string) (string, error) {
+	request := ociCore.GetSubnetRequest{
+		SubnetId: resourceID,
+	}
+
+	response, err := e.Networking.GetSubnet(context.Background(), request)
+	if err != nil {
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("subnet not found: %s", *resourceID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.Subnet.LifecycleState), nil
+}
+
+func (e *Environ) createSubnet(controllerUUID, ad, cidr string, vcnID *string, seclists []string, routeRableID *string) (ociCore.Subnet, error) {
+	displayName := fmt.Sprintf("juju-%s-%s", ad, controllerUUID)
+	compartment := e.ecfg().compartmentID()
+	// TODO(gsamfira): maybe "local" would be better?
+	subnetDetails := ociCore.CreateSubnetDetails{
+		AvailabilityDomain: &ad,
+		CidrBlock:          &cidr,
+		CompartmentId:      compartment,
+		VcnId:              vcnID,
+		DisplayName:        &displayName,
+		RouteTableId:       routeRableID,
+		SecurityListIds:    seclists,
+		// DnsLabel:           &providerNetwork.DnsLabel,
```

Comment:
This is something that we need to decide. This value, together with the hostname set on the vnic and the DNS label of the VCN, will form a FQDN. The DNS label must be unique within the VCN. Allowing the user to set this is not ideal, as he would have to create a sort of map for each availability domain. The number of availability domains may be increased in the future. The user would also have to configure the dns label of the VCN separately.

Generating them automatically may yield in FQDNs that are of no use to anyone (UUIDs?). 

So any thoughts on this are welcome.

Reply:
It feels like either the user is setting what they want the DNS labels to be, because they have meaning that they want to ascribe, or it ends up as just a UUID as you mention. So either it is something that we give the user a way to control, or we just don't set it at all.
We *could* do something like model-uuid by default, but have it configurable?